helios.js

```diff
@@ -28652,6 +28709,98 @@ class InlineDatum extends Datum {
 	}
 }
 
+/**
+ * Encoding for Transcation Matadatum
+ */
+export class MetadataString extends CborData {
+	/** @type {number[]} */
+	#bytes;
+
+	/**
+	 * @param {number[]} bytes
+	 */
+	constructor(bytes) {
+		super();
+		this.#bytes = bytes;
+	}
+
+	/**
+	 * Applies utf-8 encoding
+	 * @param {string} s
+	 * @returns {MetadataString}
+	 */
+	static fromString(s) {
+		let bytes = stringToBytes(s);
+
+		return new MetadataString(bytes);
+	}
+
+	get bytes() {
+		return this.#bytes.slice();
+	}
+
+	/**
+	 * @returns {string}
+	 */
+	toHex() {
+		return bytesToHex(this.#bytes);
+	}
+
+	toString() {
+		return `#${this.toHex()}`;
+	}
+
+	/**
+	 * @returns {number[]}
+	 */
+	toCbor() {
+		// If the data is >64 bytes then split into List
+
+		if (this.#bytes.length > 64) {
+			const chunks = [];
+			for (let i = 0; i < this.#bytes.length; i = i + 64) {
+				chunks.push(CborData.encodeUtf8(this.#bytes.slice(i, i + 64)));
+			}
+			return CborData.encodeDefList(chunks);
+		}
+
+		return CborData.encodeUtf8(this.#bytes);
+	}
+}
+
+class TxMetadata {
+	/**@type {Record<number, number[]>} */
+	#metadata;
+
+	constructor() {
+		this.#metadata = {};
+	}
+
+	/**
+	 *
+	 * @param {number} tag
+	 * @param {MetadataString} data
+	 */
+	add(tag, data) {
+		this.#metadata[tag] = data.toCbor();
+	}
+
+	dump() {
+		return Object.keys(this.#metadata).map((key) => ({
+			[key]: bytesToHex(this.#metadata[key]),
+		}))
+	}
+
+	toCbor() {
+		//  [CborData | number[], CborData | number[]][]
+		const data = Object.keys(this.#metadata).map((key) => [
+			CborData.encodeInteger(BigInt(key)),
+			this.#metadata[key],
+		]);
+		// @ts-ignore - FIXME: I don't know how to make this happy yet!
+		return CborData.encodeMap(data);
```

Comment:
You quoted https://github.com/input-output-hk/cardano-ledger/blob/master/eras/babbage/test-suite/cddl-files/babbage.cddl#L344 earlier, but I think returning the plain map like this is good enough (all node implementations should be backward compatible with shelley)

Reply:
You are correct; I've implemented the old version here! Yes it should be compatible but perhaps I will look at how this could be extended to fully support Alonzo +